tools/setup_helpers/ninja_builder.py

```diff
@@ -54,10 +54,168 @@ def patch(obj, attr_name, val):
             finally:
                 setattr(obj, attr_name, orig_val)
 
-        orig_compile = distutils.unixccompiler.UnixCCompiler._compile
-        orig_link = distutils.unixccompiler.UnixCCompiler.link
+        if os.sys.platform == 'win32':
+            orig_compile = distutils.msvccompiler.MSVCCompiler.compile
+            orig_link = distutils.msvccompiler.MSVCCompiler.link
+        else:
+            orig_compile = distutils.unixccompiler.UnixCCompiler._compile
+            orig_link = distutils.unixccompiler.UnixCCompiler.link
 
-        def _compile(self, obj, src, ext, cc_args, extra_postargs, pp_opts):
+        def win_compile(self, sources,
```

Comment:
This looks like it was copy-pasted from the stock Windows compilation logic. What makes it difficult to reuse the original function via monkey patching (similar to what is done in unix)?

Reply:
It can't be done like this. If you look into `disutils`, the class `MSVCCompiler` overrides the `compile` function rather than `_compile`, unlike what's done in `UnixCompiler`.